### DISCUSSION TITLE
Expected Loss Method

### DISCUSSION THREAD

**Discussion by henrydingliu** [casact/chainladder-python#329] [General]
noticed this method was missing from the API reference.
https://chainladder-python.readthedocs.io/en/latest/modules/api.html#module-chainladder.methods

is it possible to do the following all within the cl package, without resorting to pandas manipulation?
1. get normal cl ultimate
2. derive ultimate loss ratio
3. for latest 3 origins, set final ultimate loss ratio to average of prior 4 origins

**Comment by kennethshsu** [comment]
I think the doc could use another round of revamp, there are some known issues outstanding, for example codes not compiling even though the source codes are fine.

Anyways...

Yes, 1 and 2 are relatively easy.
```python
xyz = cl.load_sample("xyz")
cl_mod = cl.Chainladder().fit(xyz["Incurred"])
cl_mod.ultimate_ / xyz["Premium"].latest_diagonal
```

To do 3. We'll need to get the average prior LRs first, this can only be done with the help of pandas:
```python
elr = (
    (cl_mod.ultimate_ / xyz["Premium"].latest_diagonal)
    .to_frame(origin_as_datetime=True)
    .iloc[4:-3]
    .mean()
)
elr.iloc[0]
```

Fit a new model using expected loss ratio:
```python
el_mod = cl.ExpectedLoss(apriori=elr.iloc[0]).fit(
    xyz["Incurred"], sample_weight=xyz["Premium"].latest_diagonal
)
el_mod.ultimate_ 
```

Then, use `cl.VotingChainladder()`:
```python
model_weights = np.array([[1, 0]] * 8 + [[0, 1]] * 3)
vot_mod = cl.VotingChainladder(
    estimators=[("cl", cl_mod), ("el", el_mod)], weights=model_weights
).fit(xyz["Incurred"], sample_weight=xyz["Premium"].latest_diagonal)
vot_mod.ultimate_
```

**Reply by henrydingliu** [reply]
it's a shame that 3 has to be done outside of cl, when you consider that most of cl is just an organized way of taking averages of ratios. 

would it make sense to expand the EL/BF/BT methods? something like, add an n_period parameter to the class. by default n_period = 0, we go to the current logic with the single apriori. n_period = 1,2, 3, ,,,n, -1, the average ratio against the sample weight is taken and applied to the sample weight.

**Reply by kennethshsu** [reply]
In this case, I don't think it makes sense to implement additional hyper parameters for the CL/BF models. If you think about the actual actuarial models, those parameters are not required or used. The CL model, is, just the CL model - it does not depend on how many "n"s to evaluate. This is why VotingChainLadder(...) was developed, so that we can mix and match models with varying weights.

If you look at my code above a bit more carefully, the only place that you would go outside of the chainladder package is to get the a priori loss ratio, if you have this else where (say judgmentally selected), you wouldn't have to go to pandas.

I'm curious to hear why you think exiting chainladder and using pandas is not convenient?

**Comment by henrydingliu** [comment]
> If you think about the actual actuarial models, those parameters are not required or used.

same could be said about the age-to-age averages. you could always judgmentally select them.

> If you look at my code above a bit more carefully, the only place that you would go outside of the chainladder package is to get the a priori loss ratio,

still a piece of code outside of the cl pipelines

> I'm curious to hear why you think exiting chainladder and using pandas is not convenient?

not meta enough. sklearn doesn't make ppl exit and do preprocessing in np.

**Reply by kennethshsu** [reply]
Ok, if you insist on not leaving pandas, you can also use this to get the average of the loss ratios. But really, underneath, the calculations are done in pandas.

```python
(cl_mod.ultimate_ / xyz["Premium"].latest_diagonal).iloc[:, :, 4:-3, -1].mean()
```
Is this one extra line of code really that troubleing? With this, you can actually know what the assumed LR is for the most recent origin periods.

But anyway, I don't think it makes sense to have additional hyperparameters in these models. Like I mentioned, the models themselves do not require these parameters. 

@jbogaardt your thoughts on this?

**Reply by jbogaardt** [reply]
I'd have done this all in chainladder with the last line as you wrote it:
```python
(cl_mod.ultimate_ / xyz["Premium"].latest_diagonal).iloc[:, :, 4:-3, -1].mean()
```
The flexibility of the `Triangle` allows adhoc calcs like this. Does it warrant its own dedicated hyperparameter or function? I don't know. 
 That said, when it comes to building `chainladder-python` models, I think it is a reasonable goal to define a complete model within a Pipeline and this one doesn't fit in nicely. This is really a preprocessing step to compute an Apriori, and that is something that is not currently supported.

For example, it would be nice to create a `pipe` that can encapsulate this logic in one single model:
```python
pipe = cl.Pipeline(
   steps=[...]  # Can this support apriori-preprocessing?
)

pipe.fit(X, sample_weight=sample_weight)
```
It would be good to determine what the Pipeline API and any additional preprocessing estimators would look like to support something like this.

**Reply by henrydingliu** [reply]
exactly. if it was implemented as a hyperparameter, we'd then be able to properly send it into gridsearch etc.

**Reply by jbogaardt** [reply]
The only thing I'd add is that (in my experience) composition is preferable to configuration. That is, I'd rather come up with a suite of preprocessing estimators that can be composed in any number of Pipelines rather than overload the existing estimators with too many bespoke hyperparameters.

In my experience, composition in this way leads to a much more stable API that is easier to maintain and easier to use without having to constantly refer to documentation.

**Reply by henrydingliu** [reply]
"overloaded with bespoke hyperparameters", eh? is apriori really a hyperparameter or just a parameter? if it is just a parameter, is the bk class of estimators in any danger of being overloaded with hyperparameter if we are just adding the first one? and is n_period really that bespoke if it already exists in the dev transformer? hard questions no doubt.

sure, composition. i suppose there's nothing stopping anyone from loading up a static triangle using chainladder ultimates, feed that into a incrementaladditive transformer, then extract the ratio from the first dev period. now, let me refer to the documentation for that incrementaladditive transformer to figure out how to make this work...

**Reply by kennethshsu** [reply]
Wait maybe I didn't understand you correctly.

Can you explain again what you are trying to accomplish?

**Reply by henrydingliu** [reply]
```
genins = cl.load_sample("genins")
genins_dev = cl.Development().fit_transform(genins)
genins_el = cl.ExpectedLoss(n_period = 4).fit(genins_dev)
```

**Reply by henrydingliu** [reply]
actually, i've been thinking about this all wrong. rather than mudding up bk, it would make more sense to generalize the generalized cape cod method to allow arbitrary specification of origin weights, with defaults to generalized cape cod. looking through the capecod source code, is it as simple as exposing weighted_exposure?

having said that, in the spirit of building flexible pipelines, maybe capecod shouldn't be an ibnr estimator at all, but an apriori estimator that combos with BF. an apriori estimator can produce a single apriori ratio estimate, or can predict ultimate for future periods if exposures for future periods is passed in.

**Reply by henrydingliu** [reply]
and if we are already at that point, just need to break out the trend parameter into a trend and a rate change vector, and chainladder can do indications

**Reply by jbogaardt** [reply]
This makes sense to me, we just need to think through the technical design.
